### PR TITLE
Add Spring Boot integration.

### PR DESCRIPTION
**Pull request by bsideup**
By implementing `SpringApplicationRunListener`, we can install the agent
very early in the lifecycle (`starting` phase).

/cc @OlegDokuka @philwebb @sbrannen @sdeleuze @ryanjbaxter @spencergibb

### CODE REVIEW

File: spring/src/main/java/reactor/blockhound/spring/BlockHoundRunListener.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import org.springframework.core.Ordered;
+import org.springframework.core.env.ConfigurableEnvironment;
+import reactor.BlockHound;
+import reactor.blockhound.integration.BlockHoundIntegration;
+
+/**
+ * This {@link SpringApplicationRunListener} installs the agent when an app is starting
+ * by calling {@link BlockHound#install(BlockHoundIntegration...)}.
+ */
+public class BlockHoundRunListener implements SpringApplicationRunListener, Ordered {
```
Comment (bsideup) [inline spring/src/main/java/reactor/blockhound/spring/BlockHoundRunListener.java]: @philwebb does it have to be public, or I can make it internal?

File: spring/src/main/java/reactor/blockhound/spring/BlockHoundRunListener.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+    public BlockHoundRunListener(SpringApplication application, String[] args) {
+
+    }
+
+    @Override
+    public int getOrder() {
+        return Ordered.LOWEST_PRECEDENCE;
```
Comment (bsideup) [inline spring/src/main/java/reactor/blockhound/spring/BlockHoundRunListener.java]: @philwebb is it correct to use `LOWEST` here?
Wanted: "be one of the very first listeners applied"

File: spring/src/main/java/reactor/blockhound/spring/BlockHoundRunListener.java (hunk 1 above)
Comment (sbrannen) [reply spring/src/main/java/reactor/blockhound/spring/BlockHoundRunListener.java]: @philwebb, has the Boot team considered making the methods in `SpringApplicationRunListener` interface `default` methods now that Java 8 is the baseline?

File: spring/src/main/java/reactor/blockhound/spring/BlockHoundRunListener.java (hunk 2 above)
Comment (sbrannen) [reply spring/src/main/java/reactor/blockhound/spring/BlockHoundRunListener.java]: > Wanted: "be one of the very first listeners applied"

@bsideup, you'd then want to use `HIGHEST_PRECEDENCE` which has the lowest numerical value.

`org.springframework.boot.SpringApplication.getSpringFactoriesInstances(Class<T>, Class<?>[], Object...)` sorts the listeners using `org.springframework.core.annotation.AnnotationAwareOrderComparator.sort(List<?>)`.

Comment (bsideup) [reply spring/src/main/java/reactor/blockhound/spring/BlockHoundRunListener.java]: Thank you, Sam! TBH I always make this mistake :(

File: spring/src/main/java/reactor/blockhound/spring/BlockHoundRunListener.java (hunk 1 above)
Comment (philwebb) [reply spring/src/main/java/reactor/blockhound/spring/BlockHoundRunListener.java]: Good idea, will add an issue. That API is somewhat unusual as it mainly exists to break a package tangle.

Comment (sbrannen) [reply spring/src/main/java/reactor/blockhound/spring/BlockHoundRunListener.java]: > Good idea, will add an issue. 

Thanks

Comment (bsideup) [reply spring/src/main/java/reactor/blockhound/spring/BlockHoundRunListener.java]: @philwebb anything to say about the visibility of this class? 😊

**Comment by bsideup** [comment]
I'm going to close this PR since we have a more generic JUnit Platform integration now (see #13)

**Pull request closed without merging**